### PR TITLE
[go_router] Add GoRouterState parameters to GoRouterData and rename replace methods

### PR DESCRIPTION
**Pull request by ValentinVignal**
Resolves https://github.com/flutter/flutter/issues/115955

After this is merged, I'll make another PR on [go_router_builder] to replace all the references of the deprecated `redirect` to `redirectWithState`

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/go_router/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+## 5.3.0
+
+- Adds `redirectWithState` to `GoRouteData`.
+- `GoRouteData.redirect` is now deprecated in favor of `GoRouteData.redirectWithState`.
```
[the remark this answers is not in the document]
Comment (ValentinVignal) [reply packages/go_router/CHANGELOG.md]: I've updated it in [✏️ Better changelog](https://github.com/flutter/packages/pull/2848/commits/396264fb5e45337a854a4f88514995ea21fb276c)

**Comment by johnpryan** [comment]
Is it possible to keep the name `redirect`? Why are we shying away from a breaking change here?

**Comment by chunhtai** [comment]
I agree with @johnpryan that we should probably just break it, but I am not sure if we can merge this without breaking the go_router_builder's build. It we can't we would need some sort of soft transition.

**Comment by ValentinVignal** [comment]
@chunhtai @johnpryan Sure I can make it a breaking change.
In that case,
- Should I remove `buildPage` (deprecated) and rename `buildPageWithState` into `buildPage`?
- Should I also add the parameter `state` to `build`?

**Comment by chunhtai** [comment]
> Should I remove buildPage (deprecated) and rename buildPageWithState into buildPage?
We should definitrly remove buildPage, but I am not sure if we should rename it back, it feels it defeat the purpose of the @deprecated.

> Should I also add the parameter state to build?
yes I think we may as well do all of these at once.

**Comment by chunhtai** [comment]
@ValentinVignal you can make a copy of [flutter.dev/go/template](https://flutter.dev/go/template) and follow the instruction above. once the doc is available through flutter.dev/go, you can then link the flutter.dev/go link to the go_router readme as part of this pr

**Comment by chunhtai** [comment]
you can also use something like kodify to format code block

**Comment by chunhtai** [comment]
also since you are doing it now, might as well consider including renaming in https://github.com/flutter/packages/pull/2846

**Comment by ValentinVignal** [comment]
@chunhtai I've create a migration guide (see https://github.com/flutter/website/pull/7910)

**Review by chunhtai — APPROVED** [review]
LGTM, can you add the migration guide to the readme?

**Review by johnpryan — APPROVED** [review]

**Comment by ValentinVignal** [comment]
> LGTM, can you add the migration guide to the readme?

Sorry I forgot to do that.
I've included the link in the README in [📝 Add migration guide's link to the read me](https://github.com/flutter/packages/pull/2848/commits/6e6afbfa3d6c75a3d2a6021e1b0193012623b726)

**Review by johnpryan — APPROVED** [review]
LGTM

**Pull request merged**